internal/repositories/devicedefinition/devicedefinition.go

```diff
@@ -0,0 +1,256 @@
+package devicedefinition
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	gmodel "github.com/DIMO-Network/identity-api/graph/model"
+	"github.com/DIMO-Network/identity-api/internal/helpers"
+	"github.com/DIMO-Network/identity-api/internal/repositories/base"
+	"github.com/DIMO-Network/identity-api/internal/services"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	"golang.org/x/exp/slices"
+)
+
+// TokenPrefix is the prefix for a global token id for Device Definition.
+const TokenPrefix = "DD"
+
+type DeviceDefinitionTablelandCountModel struct {
+	Count int `json:"count(*)"`
+}
+
+type DeviceDefinitionTablelandModel struct {
+	Index      int    `json:"index"`
+	ID         string `json:"id"`
+	KSUID      string `json:"ksuid"`
+	Model      string `json:"model"`
+	Year       int    `json:"year"`
+	DeviceType string `json:"devicetype"`
+	ImageURI   string `json:"imageuri"`
+	Metadata   struct {
+		DeviceAttributes []struct {
+			Name  string `json:"name"`
+			Value string `json:"value"`
+		} `json:"device_attributes"`
+	} `json:"metadata"`
+}
+
+type Repository struct {
+	*base.Repository
+	ManufacturerContractService *services.ManufacturerContractService
+	TablelandApiService         *services.TablelandApiService
+	ManufacturerCacheService    *services.ManufacturerCacheService
+}
+
+func ToAPI(v *DeviceDefinitionTablelandModel) (*gmodel.DeviceDefinition, error) {
+	var result = gmodel.DeviceDefinition{}
+
+	globalID, err := base.EncodeGlobalTokenID(TokenPrefix, v.Index)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding device definition id: %w", err)
+	}
+
+	result.ID = globalID
+	result.DeviceDefinitionID = &v.ID
+	result.LegacyID = &v.KSUID
+	result.Year = &v.Year
+	result.Model = &v.Model
+	result.DeviceType = &v.DeviceType
+	result.ImageURI = &v.ImageURI
+
+	for _, attr := range v.Metadata.DeviceAttributes {
+		result.Attributes = append(result.Attributes, &gmodel.DeviceDefinitionAttribute{
+			Name:  &attr.Name,
+			Value: &attr.Value,
+		})
+	}
+
+	return &result, nil
+}
+
+func (r *Repository) GetDeviceDefinition(ctx context.Context, by gmodel.DeviceDefinitionBy) (*gmodel.DeviceDefinition, error) {
+	if len(by.ID) == 0 {
+		return nil, gqlerror.Errorf("Provide exactly one `ID`.")
+	}
+	splitID := strings.Split(by.ID, "_")
+
+	if len(splitID) < 2 {
+		return nil, gqlerror.Errorf("The `ID` is incorrect.")
+	}
+
+	manufacturerSlug := splitID[0]
+
+	manufacturers, err := r.ManufacturerCacheService.GetAllManufacturers(ctx)
```

Comment:
Why do we need to pull and cache these? We have a manufacturers table—can we store the information we need there, using chain events?

Reply:
Okay, one call you're making is to `getManufacturerIdByName`, but we already have this information in the `manufacturers` table. Let me think about the other part.

Reply:
Are we going to force them to use an empty prefix as well?